src/query_plan/operation.rs

```diff
@@ -1735,6 +2370,248 @@ pub(crate) fn merge_selection_sets(
+        fragments.iter().for_each(|(_, fragment)| {
+            let mut fragment_usages: HashMap<Name, i32> = HashMap::new();
+            NamedFragments::collect_fragment_usages(&fragment.selection_set, &mut fragment_usages);
+            let usages: Vec<Name> = fragment_usages.keys().cloned().collect::<Vec<Name>>();
+            fragments_map.insert(
+                fragment.name.clone(),
+                FragmentDependencies {
+                    fragment: fragment.clone(),
+                    depends_on: usages,
+                },
+            );
+        });
+
+        let mut removed_fragments: HashSet<Name> = HashSet::new();
+        let mut mapped_fragments = NamedFragments::default();
+        while !fragments_map.is_empty() {
+            // Note that graphQL specifies that named fragments cannot have cycles (https://spec.graphql.org/draft/#sec-Fragment-spreads-must-not-form-cycles)
+            // and so we're guaranteed that on every iteration, at least one element of the map is removed (so the `while` loop will terminate).
+            fragments_map.retain(|name, info| {
+                let can_remove = info
+                    .depends_on
+                    .iter()
+                    .all(|n| mapped_fragments.contains(n) || removed_fragments.contains(n));
+                if can_remove {
+                    if let Ok(normalized) =
+                        NormalizedFragment::normalize(&info.fragment, &mapped_fragments, schema)
+                    {
+                        // TODO this actually throws in JS code -> should we also throw?
+                        // JS code has methods for
+                        // * add and throw exception if entry already there
+                        // * add_if_not_exists
+                        // Rust HashMap exposes insert (that overwrites) and try_insert (that throws)
+                        mapped_fragments.insert(normalized);
+                    } else {
+                        removed_fragments.insert(name.clone());
+                    }
+                }
+                // keep only the elements that cannot be removed
+                !can_remove
+            });
+        }
+        mapped_fragments
+    }
+
+    // JS PORT - we need to calculate those for both SelectionSet and NormalizedSelectionSet
+    fn collect_fragment_usages(selection_set: &SelectionSet, aggregator: &mut HashMap<Name, i32>) {
+        selection_set.selections.iter().for_each(|s| match s {
+            Selection::Field(f) => {
+                NamedFragments::collect_fragment_usages(&f.selection_set, aggregator);
+            }
+            Selection::InlineFragment(i) => {
+                NamedFragments::collect_fragment_usages(&i.selection_set, aggregator);
+            }
+            Selection::FragmentSpread(f) => {
+                let current_count = aggregator.entry(f.fragment_name.clone()).or_default();
+                *current_count += 1;
+            }
```

Comment:
Should this also collect fragments used by the definition of the relevant named fragment?

Reply:
I only ported functions used by the rebasing logic -> we **DO** need to capture used fragment names when re-applying the fragments onto the operation (`Operation::optimise`). IMHO that should be tackled when working on that logic.

Reply:
in [`map_in_dependency_order`](https://github.com/apollographql/federation-next/blob/3016252afcd8c80d7e0a1b164d8563f248001044/src/query_plan/operation.rs#L2466-L2477) we iterate over selection sets of all fragment definitions to capture the references. Which is then used to `normalize` the fragments in correct order so they can reference other fragments.